File: sdk/typescript/src/did.ts
import { KYCVaultClient, DIDDocument, DIDVerificationMethod, ServiceEndpoint } from "./client";
import { DIDError } from "./errors";

export interface CreateDIDOptions {
  keyType?: string;
  network?: string;
  services?: ServiceEndpoint[];
}

export interface RotateKeyOptions {
  keyType?: string;
}

export class DIDManager {
  private readonly client: KYCVaultClient;

  constructor(client: KYCVaultClient) {
    this.client = client;
  }

  async resolve(did: string): Promise<DIDDocument> {
    this.validateDID(did);
    return this.client.resolveDID(did);
  }

  async create(method: string, options?: CreateDIDOptions): Promise<DIDDocument> {
    if (!method) {
      throw DIDError.validation("method", "DID method is required");
    }

    const supportedMethods = ["key", "web", "ethr", "ion", "indy"];
    if (!supportedMethods.includes(method)) {
      throw DIDError.unsupportedMethod(method);
    }

    return this.client.createDID(method, options);
  }

  async deactivate(did: string): Promise<void> {
    this.validateDID(did);
    await this.client.deactivateDID(did);
  }

  async rotateKey(did: string, keyId: string, options?: RotateKeyOptions): Promise<DIDDocument> {
    this.validateDID(did);
    if (!keyId) {
      throw DIDError.validation("keyId", "Key ID is required for rotation");
    }

    return this.client.rotateKey(did, keyId, options);
  }

  async resolveWithMetadata(
    did: string,
  ): Promise<{ document: DIDDocument; duration: number; resolvedAt: number }> {
    const start = performance.now();
    const doc = await this.resolve(did);
    const duration = performance.now() - start;

    return {
      document: doc,
      duration,
      resolvedAt: Math.floor(Date.now() / 1000),
    };
  }

  async getVerificationMethods(did: string): Promise<DIDVerificationMethod[]> {
    const doc = await this.resolve(did);
    return doc.verificationMethod || [];
  }

  async getServices(did: string): Promise<ServiceEndpoint[]> {
    const doc = await this.resolve(did);
    return doc.service || [];
  }

  async findServiceByType(
    did: string,
    serviceType: string,
  ): Promise<ServiceEndpoint | undefined> {
    const services = await this.getServices(did);
    return services.find((s) => s.type === serviceType);
  }

  async getController(did: string): Promise<string[]> {
    const doc = await this.resolve(did);
    return doc.controller || [doc.id];
  }

  async isDeactivated(did: string): Promise<boolean> {
    try {
      await this.resolve(did);
      return false;
    } catch (error) {
      if (error instanceof DIDError && error.message.includes("deactivated")) {
        return true;
      }
      throw error;
    }
  }

  async batchResolve(dids: string[]): Promise<Map<string, DIDDocument>> {
    const results = new Map<string, DIDDocument>();
    const entries = await Promise.allSettled(
      dids.map(async (did) => {
        const doc = await this.resolve(did);
        return { did, doc };
      }),
    );

    for (const entry of entries) {
      if (entry.status === "fulfilled") {
        results.set(entry.value.did, entry.value.doc);
      }
    }

    return results;
  }

  private validateDID(did: string): void {
    if (!did || typeof did !== "string") {
      throw DIDError.validation("did", "DID string is required");
    }

    const parts = did.split(":");
    if (parts.length < 3 || parts[0] !== "did") {
      throw DIDError.validation("did", `Invalid DID format: '${did}'`);
    }

    const method = parts[1];
    if (!method || method.length === 0) {
      throw DIDError.validation("did", "DID method is required");
    }
  }

  static parseDID(did: string): { method: string; methodSpecificId: string } {
    const parts = did.split(":");
    if (parts.length < 3 || parts[0] !== "did") {
      throw new Error(`Invalid DID: ${did}`);
    }
    return {
      method: parts[1],
      methodSpecificId: parts.slice(2).join(":"),
    };
  }

  static isValidDID(did: string): boolean {
    try {
      DIDManager.parseDID(did);
      return true;
    } catch {
      return false;
    }
  }

  static generateDocURI(did: string, fragment: string): string {
    return `${did}#${fragment}`;
  }

  static createKeyDocURI(did: string, keyTag: string): string {
    return `${did}#${keyTag}`;
  }
}

export function createDIDDocument(
  id: string,
  options?: {
    controller?: string[];
    verificationMethods?: DIDVerificationMethod[];
    services?: ServiceEndpoint[];
    alsoKnownAs?: string[];
  },
): DIDDocument {
  const doc: DIDDocument = {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
    ],
    id,
    verificationMethod: options?.verificationMethods || [],
    authentication: (options?.verificationMethods || []).map((vm) => vm.id),
    service: options?.services,
    alsoKnownAs: options?.alsoKnownAs,
    controller: options?.controller,
    created: new Date().toISOString(),
    updated: new Date().toISOString(),
  };

  return doc;
}

export function createVerificationMethod(
  id: string,
  type: string,
  controller: string,
  publicKey: { multibase?: string; jwk?: Record<string, unknown> },
): DIDVerificationMethod {
  return {
    id,
    type,
    controller,
    publicKeyMultibase: publicKey.multibase,
    publicKeyJwk: publicKey.jwk,
  };
}

export function addVerificationMethod(
  doc: DIDDocument,
  method: DIDVerificationMethod,
): DIDDocument {
  return {
    ...doc,
    verificationMethod: [...(doc.verificationMethod || []), method],
    authentication: [...(doc.authentication || []), method.id],
  };
}

export function removeVerificationMethod(
  doc: DIDDocument,
  methodId: string,
): DIDDocument {
  return {
    ...doc,
    verificationMethod: (doc.verificationMethod || []).filter((vm) => vm.id !== methodId),
    authentication: (doc.authentication || []).filter((id) => id !== methodId),
    assertionMethod: (doc.assertionMethod || []).filter((id) => id !== methodId),
    keyAgreement: (doc.keyAgreement || []).filter((id) => id !== methodId),
  };
}

export function addService(doc: DIDDocument, service: ServiceEndpoint): DIDDocument {
  return {
    ...doc,
    service: [...(doc.service || []), service],
  };
}

export function removeService(doc: DIDDocument, serviceId: string): DIDDocument {
  return {
    ...doc,
    service: (doc.service || []).filter((s) => s.id !== serviceId),
  };
}

File: sdk/flutter/lib/did.dart
import 'kyc_vault.dart';

class DIDManager {
  final KYCVaultClient _client;

  DIDManager(this._client);

  static const _supportedMethods = [
    'key',
    'web',
    'ethr',
    'ion',
    'indy',
  ];

  void _validateDID(String did) {
    if (did.isEmpty) {
      throw DIDError('DID string is required', 'VALIDATION_ERROR');
    }
    final parts = did.split(':');
    if (parts.length < 3 || parts[0] != 'did') {
      throw DIDError("Invalid DID format: '$did'", 'VALIDATION_ERROR');
    }
    if (parts[1].isEmpty) {
      throw DIDError('DID method is required', 'VALIDATION_ERROR');
    }
  }

  Future<Map<String, dynamic>> resolve(String did) async {
    _validateDID(did);
    return _client.resolveDID(did);
  }

  Future<Map<String, dynamic>> create(
    String method, {
    String? keyType,
    String? network,
    List<Map<String, dynamic>>? services,
  }) async {
    if (method.isEmpty) {
      throw DIDError('method', 'DID method is required');
    }
    if (!_supportedMethods.contains(method)) {
      throw DIDError('Unsupported DID method: $method', 'UNSUPPORTED_DID_METHOD');
    }
    return _client.createDID(method,
        keyType: keyType, network: network, services: services);
  }

  Future<void> deactivate(String did) async {
    _validateDID(did);
    await _client.deactivateDID(did);
  }

  Future<Map<String, dynamic>> rotateKey(
    String did,
    String keyId, {
    String? keyType,
  }) async {
    _validateDID(did);
    if (keyId.isEmpty) {
      throw DIDError('keyId', 'Key ID is required for rotation');
    }
    return _client.rotateKey(did, keyId, keyType: keyType);
  }

  Future<Map<String, dynamic>> resolveWithMetadata(String did) async {
    final start = DateTime.now().millisecondsSinceEpoch;
    final document = await resolve(did);
    final duration = DateTime.now().millisecondsSinceEpoch - start;
    return {
      'document': document,
      'duration': duration,
      'resolvedAt': DateTime.now().millisecondsSinceEpoch ~/ 1000,
    };
  }

  Future<List<Map<String, dynamic>>> getVerificationMethods(String did) async {
    final doc = await resolve(did);
    return List<Map<String, dynamic>>.from(doc['verificationMethod'] ?? []);
  }

  Future<List<Map<String, dynamic>>> getServices(String did) async {
    final doc = await resolve(did);
    return List<Map<String, dynamic>>.from(doc['service'] ?? []);
  }

  Future<Map<String, dynamic>?> findServiceByType(
    String did,
    String serviceType,
  ) async {
    final services = await getServices(did);
    try {
      return services.firstWhere((s) => s['type'] == serviceType);
    } catch (_) {
      return null;
    }
  }

  Future<List<String>> getController(String did) async {
    final doc = await resolve(did);
    final controller = doc['controller'];
    if (controller != null) {
      return List<String>.from(controller);
    }
    return [doc['id'] as String];
  }

  Future<bool> isDeactivated(String did) async {
    try {
      await resolve(did);
      return false;
    } catch (e) {
      if (e is DIDError && e.toString().contains('deactivated')) {
        return true;
      }
      rethrow;
    }
  }

  Future<Map<String, Map<String, dynamic>>> batchResolve(
    List<String> dids,
  ) async {
    final results = <String, Map<String, dynamic>>{};
    final futures = dids.map((did) async {
      try {
        final doc = await resolve(did);
        return MapEntry(did, doc);
      } catch (_) {
        return null;
      }
    });
    final entries = await Future.wait(futures);
    for (final entry in entries) {
      if (entry != null) {
        results[entry.key] = entry.value;
      }
    }
    return results;
  }

  static Map<String, String> parseDID(String did) {
    final parts = did.split(':');
    if (parts.length < 3 || parts[0] != 'did') {
      throw Exception('Invalid DID: $did');
    }
    return {
      'method': parts[1],
      'methodSpecificId': parts.sublist(2).join(':'),
    };
  }

  static bool isValidDID(String did) {
    try {
      parseDID(did);
      return true;
    } catch (_) {
      return false;
    }
  }

  static String generateDocURI(String did, String fragment) => '$did#$fragment';
  static String createKeyDocURI(String did, String keyTag) => '$did#$keyTag';

  Map<String, dynamic> createDIDDocument(
    String id, {
    List<String>? controller,
    List<Map<String, dynamic>>? verificationMethods,
    List<Map<String, dynamic>>? services,
    List<String>? alsoKnownAs,
  }) {
    return {
      '@context': [
        'https://www.w3.org/ns/did/v1',
        'https://w3id.org/security/suites/ed25519-2020/v1',
      ],
      'id': id,
      'verificationMethod': verificationMethods ?? [],
      'authentication':
          (verificationMethods ?? []).map((vm) => vm['id']).toList(),
      'service': services,
      'alsoKnownAs': alsoKnownAs,
      'controller': controller,
      'created': DateTime.now().toIso8601String(),
      'updated': DateTime.now().toIso8601String(),
    };
  }

  Map<String, dynamic> createVerificationMethod(
    String id,
    String type,
    String controller,
    Map<String, dynamic> publicKey,
  ) {
    return {
      'id': id,
      'type': type,
      'controller': controller,
      'publicKeyMultibase': publicKey['multibase'],
      'publicKeyJwk': publicKey['jwk'],
    };
  }

  Map<String, dynamic> addVerificationMethod(
    Map<String, dynamic> doc,
    Map<String, dynamic> method,
  ) {
    final methods = List<Map<String, dynamic>>.from(doc['verificationMethod'] ?? []);
    final auth = List<String>.from(doc['authentication'] ?? []);
    return {
      ...doc,
      'verificationMethod': [...methods, method],
      'authentication': [...auth, method['id']],
    };
  }

  Map<String, dynamic> removeVerificationMethod(
    Map<String, dynamic> doc,
    String methodId,
  ) {
    return {
      ...doc,
      'verificationMethod': (doc['verificationMethod'] as List?)
              ?.where((vm) => vm['id'] != methodId)
              .toList() ??
          [],
      'authentication': (doc['authentication'] as List?)
              ?.where((id) => id != methodId)
              .toList() ??
          [],
      'assertionMethod': (doc['assertionMethod'] as List?)
              ?.where((id) => id != methodId)
              .toList() ??
          [],
      'keyAgreement': (doc['keyAgreement'] as List?)
              ?.where((id) => id != methodId)
              .toList() ??
          [],
    };
  }

  Map<String, dynamic> addService(
    Map<String, dynamic> doc,
    Map<String, dynamic> service,
  ) {
    final services = List<Map<String, dynamic>>.from(doc['service'] ?? []);
    return {
      ...doc,
      'service': [...services, service],
    };
  }

  Map<String, dynamic> removeService(
    Map<String, dynamic> doc,
    String serviceId,
  ) {
    return {
      ...doc,
      'service': (doc['service'] as List?)
              ?.where((s) => s['id'] != serviceId)
              .toList() ??
          [],
    };
  }
}

class DIDError implements Exception {
  final String message;
  final String code;
  final Map<String, dynamic>? details;

  DIDError(this.message, this.code, [this.details]);

  @override
  String toString() => 'DIDError($code): $message';
}
